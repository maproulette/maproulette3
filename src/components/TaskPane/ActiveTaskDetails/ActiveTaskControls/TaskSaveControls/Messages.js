import { defineMessages } from 'react-intl'

/**
 * Internationalized messages for use with TaskSaveControls
 */
export default defineMessages({
  save: {
    id: "Task.controls.save.label",
    defaultMessage: "Save",
  },
  unsave: {
    id: "Task.controls.unsave.label",
    defaultMessage: "Unsave",
  },
})
