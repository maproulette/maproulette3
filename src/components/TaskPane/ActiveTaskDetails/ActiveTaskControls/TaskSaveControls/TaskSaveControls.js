import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { FormattedMessage } from 'react-intl'
import _findIndex from 'lodash/findIndex'
import SvgSymbol from '../../../../SvgSymbol/SvgSymbol'
import messages from './Messages'

/**
 * TaskSaveControls displays controls for saving or unsaving the current task
 * into the user's set of saved tasks.
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export default class TaskSaveControls extends Component {
  render() {
    if (!this.props.user || !this.props.task) {
      return null
    }

    const taskIsSaved =
      _findIndex(this.props.user.savedTasks, {id: this.props.task.id}) !== -1

    let saveTaskButton = null
    let unsaveTaskButton = null

    if (taskIsSaved) {
      unsaveTaskButton = (
        <p className="control">
          <button className={classNames("button save-task-toggle unsave-task",
                                        {"large-and-wide": !this.props.isMinimized,
                                        "icon-only": this.props.isMinimized})}
                  onClick={() => this.props.unsaveTask(this.props.user.id, this.props.task.id)}>
            <span className="control-icon"
                  title={this.props.intl.formatMessage(messages.unsave)}>
              <SvgSymbol viewBox='0 0 20 20' sym="bookmark-icon" />
            </span>
            <span className="control-label">
              <FormattedMessage {...messages.unsave} />
            </span>
          </button>
        </p>
      )
    }
    else {
      saveTaskButton = (
        <p className="control">
          <button className={classNames("button save-task-toggle save-task",
                                        {"large-and-wide": !this.props.isMinimized,
                                        "icon-only": this.props.isMinimized})}
                  onClick={() => this.props.saveTask(this.props.user.id, this.props.task.id)}>
            <span className="control-icon"
                  title={this.props.intl.formatMessage(messages.save)}>
              <SvgSymbol viewBox='0 0 20 20' sym="bookmark-add-icon" />
            </span>
            <span className="control-label">
              <FormattedMessage {...messages.save} />
            </span>
          </button>
        </p>
      )
    }

    return (
      <div className={classNames("task-save-controls", this.props.className)}>
        {saveTaskButton}
        {unsaveTaskButton}
      </div>
    )
  }
}

TaskSaveControls.propTypes = {
  /** The current user */
  user: PropTypes.object,
  /** The current active task */
  task: PropTypes.object.isRequired,
  /** Invoked if the user decides to save the task */
  saveTask: PropTypes.func.isRequired,
  /** Invoked if the user decides to unsave the task */
  unsaveTask: PropTypes.func.isRequired,
}
