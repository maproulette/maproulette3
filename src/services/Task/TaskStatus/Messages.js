import { defineMessages } from 'react-intl'

/**
 * Internationalized messages for use with TaskStatus.
 */
export default defineMessages({
  created: {
    id: "Task.status.created",
    defaultMessage: "Created"
  },
  fixed: {
    id: "Task.status.fixed",
    defaultMessage: "Fixed"
  },
  falsePositive: {
    id: "Task.status.falsePositive",
    defaultMessage: "Not an Issue"
  },
  skipped: {
    id: "Task.status.skipped",
    defaultMessage: "Skipped"
  },
  deleted: {
    id: "Task.status.deleted",
    defaultMessage: "Deleted"
  },
  disabled: {
    id: "Task.status.disabled",
    defaultMessage: "Disabled"
  },
  alreadyFixed: {
    id: "Task.status.alreadyFixed",
    defaultMessage: "Already Fixed"
  },
  tooHard: {
    id: "Task.status.tooHard",
    defaultMessage: "Can't Complete"
  },
})
