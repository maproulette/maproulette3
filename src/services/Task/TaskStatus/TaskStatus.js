import _map from 'lodash/map'
import _invert from 'lodash/invert'
import _fromPairs from 'lodash/fromPairs'
import messages from './Messages'

// These statuses are defined on the server
export const TASK_STATUS_CREATED = 0
export const TASK_STATUS_FIXED = 1
export const TASK_STATUS_FALSE_POSITIVE = 2
export const TASK_STATUS_SKIPPED = 3
export const TASK_STATUS_DELETED = 4
export const TASK_STATUS_ALREADY_FIXED = 5
export const TASK_STATUS_TOO_HARD = 6

export const TaskStatus = Object.freeze({
  created: TASK_STATUS_CREATED,
  fixed: TASK_STATUS_FIXED,
  falsePositive: TASK_STATUS_FALSE_POSITIVE,
  skipped: TASK_STATUS_SKIPPED,
  deleted: TASK_STATUS_DELETED,
  alreadyFixed: TASK_STATUS_ALREADY_FIXED,
  tooHard: TASK_STATUS_TOO_HARD,
})

export const keysByStatus = Object.freeze(_invert(TaskStatus))

/**
 * Returns a Set of status progressions that are allowed
 * for the given status. An empty Set is returned if no
 * progressions are allowed.
 *
 * Set includeSelf to true if the given (presumed current) status should be
 * also included in the results.
 *
 * @returns a Set of allowed status progressions
 */
export const allowedStatusProgressions = function(status, includeSelf = false) {
  let progressions = null
  switch(status) {
    case TaskStatus.created:
      progressions = new Set([TaskStatus.fixed, TaskStatus.falsePositive,
                              TaskStatus.skipped, TaskStatus.deleted,
                              TaskStatus.alreadyFixed, TaskStatus.tooHard])
      break
    case TaskStatus.fixed:
      progressions = new Set()
      break
    case TaskStatus.falsePositive:
      progressions = new Set([TaskStatus.fixed])
      break
    case TaskStatus.skipped:
    case TaskStatus.tooHard:
      progressions = new Set([TaskStatus.fixed, TaskStatus.falsePositive,
                              TaskStatus.skipped, TaskStatus.alreadyFixed,
                              TaskStatus.tooHard])
      break
    case TaskStatus.deleted:
      progressions = new Set([TaskStatus.created])
      break
    case TaskStatus.alreadyFixed:
      progressions = new Set()
      break
    default:
      throw new Error("unrecognized-task-status",
                      `Unrecognized task status ${status}`)
  }

  if (includeSelf) {
    progressions.add(status)
  }

  return progressions
}

/**
 * Returns an object mapping difficulty values to raw internationalized
 * messages suitable for use with FormattedMessage or formatMessage.
 */
export const messagesByStatus = _fromPairs(
  _map(messages, (message, key) => [TaskStatus[key], message])
)

/** Returns object containing localized labels  */
export const statusLabels = intl => _fromPairs(
  _map(messages, (message, key) => [key, intl.formatMessage(message)])
)
