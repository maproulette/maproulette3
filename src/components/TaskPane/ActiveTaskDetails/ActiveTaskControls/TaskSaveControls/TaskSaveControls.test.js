import React from 'react'
import { omit as _omit, cloneDeep as _cloneDeep } from 'lodash'
import TaskSaveControls from './TaskSaveControls'
import { TaskStatus } from '../../../../../services/Task/TaskStatus/TaskStatus'

const propsFixture = {
  task: {
    id: 123,
    status: TaskStatus.fixed,
    parent: {
      id: 321,
    }
  },
  user: {
    id: 357,
    savedTasks: [],
    isLoggedIn: true,
  },
  intl: {formatMessage: jest.fn()},
}

let basicProps = null

beforeEach(() => {
  basicProps = _cloneDeep(propsFixture)

  basicProps.saveTask = jest.fn()
  basicProps.unsaveTask = jest.fn()
})

test("it renders a save button if the task has not been saved by the user", () => {
  const wrapper = shallow(
    <TaskSaveControls {...basicProps} />
  )

  expect(wrapper.find('.button.save-task').exists()).toBe(true)
  expect(wrapper.find('.button.unsave-task').exists()).toBe(false)

  expect(wrapper).toMatchSnapshot()
})

test("clicking the save button signals that the task should be saved", () => {
  const wrapper = shallow(
    <TaskSaveControls {...basicProps} />
  )

  wrapper.find('.button.save-task').simulate('click')

  expect(basicProps.saveTask.mock.calls.length).toBe(1)
  expect(basicProps.saveTask.mock.calls[0][0]).toBe(basicProps.user.id)
  expect(basicProps.saveTask.mock.calls[0][1]).toBe(basicProps.task.id)
})

test("it renders an unsave button if the task has already been saved by the user", () => {
  basicProps.user.savedTasks.push({id: basicProps.task.id})

  const wrapper = shallow(
    <TaskSaveControls {...basicProps} />
  )

  expect(wrapper.find('.button.unsave-task').exists()).toBe(true)
  expect(wrapper.find('.button.save-task').exists()).toBe(false)

  expect(wrapper).toMatchSnapshot()
})

test("clicking the unsave button signals that the task should be unsaved", () => {
  basicProps.user.savedTasks.push({id: basicProps.task.id})

  const wrapper = shallow(
    <TaskSaveControls {...basicProps} />
  )

  wrapper.find('.button.unsave-task').simulate('click')

  expect(basicProps.unsaveTask.mock.calls.length).toBe(1)
  expect(basicProps.unsaveTask.mock.calls[0][0]).toBe(basicProps.user.id)
  expect(basicProps.unsaveTask.mock.calls[0][1]).toBe(basicProps.task.id)
})
