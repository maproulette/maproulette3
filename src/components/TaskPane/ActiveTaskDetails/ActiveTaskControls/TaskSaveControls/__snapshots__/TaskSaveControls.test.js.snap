// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a save button if the task has not been saved by the user 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskSaveControls
    intl={
        Object {
            "formatMessage": [Function],
          }
    }
    saveTask={[Function]}
    task={
        Object {
            "id": 123,
            "parent": Object {
              "id": 321,
            },
            "status": 1,
          }
    }
    unsaveTask={[Function]}
    user={
        Object {
            "id": 357,
            "isLoggedIn": true,
            "savedTasks": Array [],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          className="control"
>
          <button
                    className="button save-task-toggle save-task large-and-wide"
                    onClick={[Function]}
          >
                    <span
                              className="control-icon"
                              title={undefined}
                    >
                              <SvgSymbol
                                        sym="bookmark-add-icon"
                                        viewBox="0 0 20 20"
                              />
                    </span>
                    <span
                              className="control-label"
                    >
                              <FormattedMessage
                                        defaultMessage="Save"
                                        id="Task.controls.save.label"
                                        values={Object {}}
                              />
                    </span>
          </button>
</p>,
        null,
      ],
      "className": "task-save-controls",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="button save-task-toggle save-task large-and-wide"
            onClick={[Function]}
>
            <span
                        className="control-icon"
                        title={undefined}
            >
                        <SvgSymbol
                                    sym="bookmark-add-icon"
                                    viewBox="0 0 20 20"
                        />
            </span>
            <span
                        className="control-label"
            >
                        <FormattedMessage
                                    defaultMessage="Save"
                                    id="Task.controls.save.label"
                                    values={Object {}}
                        />
            </span>
</button>,
          "className": "control",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="control-icon"
                title={undefined}
>
                <SvgSymbol
                                sym="bookmark-add-icon"
                                viewBox="0 0 20 20"
                />
</span>,
              <span
                className="control-label"
>
                <FormattedMessage
                                defaultMessage="Save"
                                id="Task.controls.save.label"
                                values={Object {}}
                />
</span>,
            ],
            "className": "button save-task-toggle save-task large-and-wide",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <SvgSymbol
                  sym="bookmark-add-icon"
                  viewBox="0 0 20 20"
/>,
                "className": "control-icon",
                "title": undefined,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "sym": "bookmark-add-icon",
                  "viewBox": "0 0 20 20",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <FormattedMessage
                  defaultMessage="Save"
                  id="Task.controls.save.label"
                  values={Object {}}
/>,
                "className": "control-label",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "defaultMessage": "Save",
                  "id": "Task.controls.save.label",
                  "values": Object {},
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "span",
            },
          ],
          "type": "button",
        },
        "type": "p",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="control"
>
            <button
                        className="button save-task-toggle save-task large-and-wide"
                        onClick={[Function]}
            >
                        <span
                                    className="control-icon"
                                    title={undefined}
                        >
                                    <SvgSymbol
                                                sym="bookmark-add-icon"
                                                viewBox="0 0 20 20"
                                    />
                        </span>
                        <span
                                    className="control-label"
                        >
                                    <FormattedMessage
                                                defaultMessage="Save"
                                                id="Task.controls.save.label"
                                                values={Object {}}
                                    />
                        </span>
            </button>
</p>,
          null,
        ],
        "className": "task-save-controls",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="button save-task-toggle save-task large-and-wide"
              onClick={[Function]}
>
              <span
                            className="control-icon"
                            title={undefined}
              >
                            <SvgSymbol
                                          sym="bookmark-add-icon"
                                          viewBox="0 0 20 20"
                            />
              </span>
              <span
                            className="control-label"
              >
                            <FormattedMessage
                                          defaultMessage="Save"
                                          id="Task.controls.save.label"
                                          values={Object {}}
                            />
              </span>
</button>,
            "className": "control",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span
                  className="control-icon"
                  title={undefined}
>
                  <SvgSymbol
                                    sym="bookmark-add-icon"
                                    viewBox="0 0 20 20"
                  />
</span>,
                <span
                  className="control-label"
>
                  <FormattedMessage
                                    defaultMessage="Save"
                                    id="Task.controls.save.label"
                                    values={Object {}}
                  />
</span>,
              ],
              "className": "button save-task-toggle save-task large-and-wide",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <SvgSymbol
                    sym="bookmark-add-icon"
                    viewBox="0 0 20 20"
/>,
                  "className": "control-icon",
                  "title": undefined,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "sym": "bookmark-add-icon",
                    "viewBox": "0 0 20 20",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <FormattedMessage
                    defaultMessage="Save"
                    id="Task.controls.save.label"
                    values={Object {}}
/>,
                  "className": "control-label",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "defaultMessage": "Save",
                    "id": "Task.controls.save.label",
                    "values": Object {},
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "span",
              },
            ],
            "type": "button",
          },
          "type": "p",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`it renders an unsave button if the task has already been saved by the user 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskSaveControls
    intl={
        Object {
            "formatMessage": [Function],
          }
    }
    saveTask={[Function]}
    task={
        Object {
            "id": 123,
            "parent": Object {
              "id": 321,
            },
            "status": 1,
          }
    }
    unsaveTask={[Function]}
    user={
        Object {
            "id": 357,
            "isLoggedIn": true,
            "savedTasks": Array [
              Object {
                "id": 123,
              },
            ],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        null,
        <p
          className="control"
>
          <button
                    className="button save-task-toggle unsave-task large-and-wide"
                    onClick={[Function]}
          >
                    <span
                              className="control-icon"
                              title={undefined}
                    >
                              <SvgSymbol
                                        sym="bookmark-icon"
                                        viewBox="0 0 20 20"
                              />
                    </span>
                    <span
                              className="control-label"
                    >
                              <FormattedMessage
                                        defaultMessage="Unsave"
                                        id="Task.controls.unsave.label"
                                        values={Object {}}
                              />
                    </span>
          </button>
</p>,
      ],
      "className": "task-save-controls",
    },
    "ref": null,
    "rendered": Array [
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="button save-task-toggle unsave-task large-and-wide"
            onClick={[Function]}
>
            <span
                        className="control-icon"
                        title={undefined}
            >
                        <SvgSymbol
                                    sym="bookmark-icon"
                                    viewBox="0 0 20 20"
                        />
            </span>
            <span
                        className="control-label"
            >
                        <FormattedMessage
                                    defaultMessage="Unsave"
                                    id="Task.controls.unsave.label"
                                    values={Object {}}
                        />
            </span>
</button>,
          "className": "control",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="control-icon"
                title={undefined}
>
                <SvgSymbol
                                sym="bookmark-icon"
                                viewBox="0 0 20 20"
                />
</span>,
              <span
                className="control-label"
>
                <FormattedMessage
                                defaultMessage="Unsave"
                                id="Task.controls.unsave.label"
                                values={Object {}}
                />
</span>,
            ],
            "className": "button save-task-toggle unsave-task large-and-wide",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <SvgSymbol
                  sym="bookmark-icon"
                  viewBox="0 0 20 20"
/>,
                "className": "control-icon",
                "title": undefined,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "sym": "bookmark-icon",
                  "viewBox": "0 0 20 20",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <FormattedMessage
                  defaultMessage="Unsave"
                  id="Task.controls.unsave.label"
                  values={Object {}}
/>,
                "className": "control-label",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "defaultMessage": "Unsave",
                  "id": "Task.controls.unsave.label",
                  "values": Object {},
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "span",
            },
          ],
          "type": "button",
        },
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          null,
          <p
            className="control"
>
            <button
                        className="button save-task-toggle unsave-task large-and-wide"
                        onClick={[Function]}
            >
                        <span
                                    className="control-icon"
                                    title={undefined}
                        >
                                    <SvgSymbol
                                                sym="bookmark-icon"
                                                viewBox="0 0 20 20"
                                    />
                        </span>
                        <span
                                    className="control-label"
                        >
                                    <FormattedMessage
                                                defaultMessage="Unsave"
                                                id="Task.controls.unsave.label"
                                                values={Object {}}
                                    />
                        </span>
            </button>
</p>,
        ],
        "className": "task-save-controls",
      },
      "ref": null,
      "rendered": Array [
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="button save-task-toggle unsave-task large-and-wide"
              onClick={[Function]}
>
              <span
                            className="control-icon"
                            title={undefined}
              >
                            <SvgSymbol
                                          sym="bookmark-icon"
                                          viewBox="0 0 20 20"
                            />
              </span>
              <span
                            className="control-label"
              >
                            <FormattedMessage
                                          defaultMessage="Unsave"
                                          id="Task.controls.unsave.label"
                                          values={Object {}}
                            />
              </span>
</button>,
            "className": "control",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span
                  className="control-icon"
                  title={undefined}
>
                  <SvgSymbol
                                    sym="bookmark-icon"
                                    viewBox="0 0 20 20"
                  />
</span>,
                <span
                  className="control-label"
>
                  <FormattedMessage
                                    defaultMessage="Unsave"
                                    id="Task.controls.unsave.label"
                                    values={Object {}}
                  />
</span>,
              ],
              "className": "button save-task-toggle unsave-task large-and-wide",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <SvgSymbol
                    sym="bookmark-icon"
                    viewBox="0 0 20 20"
/>,
                  "className": "control-icon",
                  "title": undefined,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "sym": "bookmark-icon",
                    "viewBox": "0 0 20 20",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <FormattedMessage
                    defaultMessage="Unsave"
                    id="Task.controls.unsave.label"
                    values={Object {}}
/>,
                  "className": "control-label",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "defaultMessage": "Unsave",
                    "id": "Task.controls.unsave.label",
                    "values": Object {},
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "span",
              },
            ],
            "type": "button",
          },
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
